[BUG] Weird but semi-consistent test failures

### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

CI

### What happened and when does this occur?

```Markdown
Not sure when it started exactly, but as of now, serial tests in `test/tests/serial/tests.bats` do not run (sometimes), because the setup procedure fails.
```


### What did you expect to happen?

```Markdown
I expected the tests to succeed as there is nothing that should prevent them from succeeding.
```


### How do we replicate the issue?

```Markdown
1. Open a PR
2. Watch how the tests for this PR fail sporadically

or

1. Merge a PR
2. Watch how the tests on `master` fail
```


### DMS version

v11.3.1

### What operating system is DMS running on?

Linux

### Which operating system version?

Ubuntu 22.04

### What instruction set architecture is DMS running on?

AMD64 / x86_64

### What container orchestration tool are you using?

Docker

### docker-compose.yml

```yml
N/A
```


### Relevant log output

```bash
not ok 212 setup_file failed
# (from function `repeat_until_success_or_timeout' in file test/test_helper/common.bash, line 43,
#  from function `wait_until_change_detection_event_completes' in file test/test_helper/common.bash, line 208,
#  from function `setup_file' in test file test/tests/serial/tests.bats, line 50)
#   `wait_until_change_detection_event_completes mail' failed
# 8d7ce2717447a5b330bc24466c6b6b3110fff825af10b80fbc0248287fa3372d
# [   INF   ]  mail.my-domain.com is up and running
# Timed out on command: __is_changedetector_processing
# mail
# bats warning: Executed 212 instead of expected 309 tests
make: *** [Makefile:52: tests/serial] Error 1
Error: Process completed with exit code 2.
```


### Other relevant information

An example failure can be seen via the following link: <https://github.com/docker-mailserver/docker-mailserver/actions/runs/3789635420/jobs/6443579882#step:5:219>

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [ ] I am rather experienced with docker
- [ ] I am inexperienced with mail servers
- [ ] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [ ] I am rather comfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

### Improvements to this form?

_No response_

--- a/test/helper/common.bash
+++ b/test/helper/common.bash
@@ -153,7 +153,7 @@ function wait_for_changes_to_be_detected_in_container() {
   repeat_in_container_until_success_or_timeout "${TIMEOUT}" "${CONTAINER_NAME}" bash -c 'source /usr/local/bin/helpers/index.sh; _obtain_hostname_and_domainname; cmp --silent -- <(_monitored_files_checksums) "${CHKSUM_FILE}" >/dev/null'
 }
 
-# Relies on ENV `LOG_LEVEL=debug` or higher
+# NOTE: Relies on ENV `LOG_LEVEL=debug` or higher
 function wait_until_change_detection_event_completes() {
   local CONTAINER_NAME="${1}"
   # Ensure early failure if arg is missing:
@@ -164,31 +164,20 @@ function wait_until_change_detection_event_completes() {
     $(docker exec "${CONTAINER_NAME}" env | grep '^LOG_LEVEL=') \
     '=(debug|trace)$'
 
-  local CHANGE_EVENT_START='Change detected'
-  local CHANGE_EVENT_END='Completed handling of detected change' # debug log
-
-  function __change_event_status() {
-    docker exec "${CONTAINER_NAME}" \
-      grep -oE "${CHANGE_EVENT_START}|${CHANGE_EVENT_END}" /var/log/supervisor/changedetector.log \
-      | tail -1
-  }
-
-  function __is_changedetector_processing() {
-    [[ $(__change_event_status) == "${CHANGE_EVENT_START}" ]]
+  # NOTE: Change events can start and finish all within < 1 sec,
+  # Reliably track the completion of a change event by comparing the before/after count:
+  function __change_event_count() {
+    docker exec "${CONTAINER_NAME}" grep --count "${CHANGE_EVENT_END}" /var/log/supervisor/changedetector.log
   }
 
   function __is_changedetector_finished() {
-    [[ $(__change_event_status) == "${CHANGE_EVENT_END}" ]]
+    [[ $(__change_event_count) -gt "${NUM_CHANGE_EVENTS_BEFORE}" ]]
   }
 
-  # A new change event is expected,
-  # If the last event status is not yet `CHANGE_EVENT_START`, wait until it is:
-  if ! __is_changedetector_processing
-  then
-    repeat_until_success_or_timeout 60 __is_changedetector_processing
-  fi
+  # Count by completions of this debug log line from `check-for-changes.sh`:
+  local CHANGE_EVENT_END='Completed handling of detected change'
+  local NUM_CHANGE_EVENTS_BEFORE=$(__change_event_count)
 
-  # Change event is in progress, wait until it finishes:
   repeat_until_success_or_timeout 60 __is_changedetector_finished
 }
 

--- a/test/test_helper/common.bash
+++ b/test/test_helper/common.bash
@@ -173,7 +173,7 @@ function wait_for_changes_to_be_detected_in_container() {
   repeat_in_container_until_success_or_timeout "${TIMEOUT}" "${CONTAINER_NAME}" bash -c 'source /usr/local/bin/helpers/index.sh; _obtain_hostname_and_domainname; cmp --silent -- <(_monitored_files_checksums) "${CHKSUM_FILE}" >/dev/null'
 }
 
-# Relies on ENV `LOG_LEVEL=debug` or higher
+# NOTE: Relies on ENV `LOG_LEVEL=debug` or higher
 function wait_until_change_detection_event_completes() {
   local CONTAINER_NAME="${1}"
   # Ensure early failure if arg is missing:
@@ -184,31 +184,20 @@ function wait_until_change_detection_event_completes() {
     $(docker exec "${CONTAINER_NAME}" env | grep '^LOG_LEVEL=') \
     '=(debug|trace)$'
 
-  local CHANGE_EVENT_START='Change detected'
-  local CHANGE_EVENT_END='Completed handling of detected change' # debug log
-
-  function __change_event_status() {
-    docker exec "${CONTAINER_NAME}" \
-      grep -oE "${CHANGE_EVENT_START}|${CHANGE_EVENT_END}" /var/log/supervisor/changedetector.log \
-      | tail -1
-  }
-
-  function __is_changedetector_processing() {
-    [[ $(__change_event_status) == "${CHANGE_EVENT_START}" ]]
+  # NOTE: Change events can start and finish all within < 1 sec,
+  # Reliably track the completion of a change event by comparing the before/after count:
+  function __change_event_count() {
+    docker exec "${CONTAINER_NAME}" grep --count "${CHANGE_EVENT_END}" /var/log/supervisor/changedetector.log
   }
 
   function __is_changedetector_finished() {
-    [[ $(__change_event_status) == "${CHANGE_EVENT_END}" ]]
+    [[ $(__change_event_count) -gt "${NUM_CHANGE_EVENTS_BEFORE}" ]]
   }
 
-  # A new change event is expected,
-  # If the last event status is not yet `CHANGE_EVENT_START`, wait until it is:
-  if ! __is_changedetector_processing
-  then
-    repeat_until_success_or_timeout 60 __is_changedetector_processing
-  fi
+  # Count by completions of this debug log line from `check-for-changes.sh`:
+  local CHANGE_EVENT_END='Completed handling of detected change'
+  local NUM_CHANGE_EVENTS_BEFORE=$(__change_event_count)
 
-  # Change event is in progress, wait until it finishes:
   repeat_until_success_or_timeout 60 __is_changedetector_finished
 }
 

